### PR TITLE
Improvements for EDN encoder in debugger

### PR DESCRIPTION
**Pull request by AGulev**
Pay attention that here we are talking about 450 000 - 500 000 only tables or refs to tables + huge amount of the other values
```
Test prj
DEBUG:SCRIPT: table_counter	180900
DEBUG:SCRIPT: ref_counter	450000

Before:
DEBUG:SCRIPT: Result time = 	3.1616051197052
DEBUG:SCRIPT: Result time = 	3.0216929912567
DEBUG:SCRIPT: Result time = 	3.0465228557587
DEBUG:SCRIPT: Result time = 	3.0130741596222
DEBUG:SCRIPT: Result time = 	3.0043928623199
DEBUG:SCRIPT: Result time = 	3.0025200843811
DEBUG:SCRIPT: Result time = 	3.0001330375671

After:
DEBUG:SCRIPT: Result time = 	1.7840061187744
DEBUG:SCRIPT: Result time = 	1.8551151752472
DEBUG:SCRIPT: Result time = 	1.8466029167175
DEBUG:SCRIPT: Result time = 	1.7579870223999
DEBUG:SCRIPT: Result time = 	1.8628711700439


Another big test project:
Before:
DEBUG:SCRIPT: Result time = 	8.8547360897064
DEBUG:SCRIPT: Result time = 	8.8110229969025
DEBUG:SCRIPT: Result time = 	8.5735511779785
DEBUG:SCRIPT: Result time = 	8.6874620914459

After:
DEBUG:SCRIPT: Result time = 	5.7142062187195	 cached_string	2056917	 encoded_strs	476642
DEBUG:SCRIPT: Result time = 	5.2303631305695	 cached_string	2533328	 encoded_strs	68
DEBUG:SCRIPT: Result time = 	5.1561901569366	 cached_string	2533404	 encoded_strs	40
DEBUG:SCRIPT: Result time = 	5.0541839599609	 cached_string	2533514	 encoded_strs	60
DEBUG:SCRIPT: Result time = 	5.0240840911865	 cached_string	2533549	 encoded_strs	34
```

(these changes shouldn't be in Release Notes)

### CODE REVIEW

File: engine/engine/content/builtins/scripts/edn.lua [hunk 1]
```diff
@@ -6,6 +6,12 @@
 
 local M = {}
 
+local tostring = tostring
+local string = string
+local debug = debug
+local type = type
+local table = table
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:13]: localisation of some frequently used functions/tables

File: engine/engine/content/builtins/scripts/edn.lua [hunk 2]
```diff
@@ -14,34 +20,15 @@ local escape_char_map = {
...
-  debug.setmetatable(val, old_mt)
-  return raw_str
-end
-
-local function has_custom_tostring(val)
-  local mt = debug.getmetatable(val)
-  return mt ~= nil and mt.__tostring ~= nil
-end
-
-local function get_prefixed_addr(prefix, val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:30]: it's faster to get address using `string.format("%p", table)`

File: engine/engine/content/builtins/scripts/edn.lua [hunk 3]
```diff
@@ -14,34 +20,15 @@ local escape_char_map = {
   ["\t" ] = "\\t",
 }
 
-local function raw_tostring(val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:17]: used only in `get_prefixed_addr`

File: engine/engine/content/builtins/scripts/edn.lua [hunk 4]
```diff
@@ -14,34 +20,15 @@ local escape_char_map = {
...
-  local old_mt = debug.getmetatable(val)
-  debug.setmetatable(val, nil)
-  local raw_str = tostring(val)
-  debug.setmetatable(val, old_mt)
-  return raw_str
-end
-
-local function has_custom_tostring(val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:25]: Merged into a function where it's used to reduce count of `getmetatable()` or `debug.getmetatable()` calls

File: engine/engine/content/builtins/scripts/edn.lua [hunk 5]
```diff
@@ -14,34 +20,15 @@ local escape_char_map = {
...
-  else
-    return ""
-  end
-end
-
 local function str(val)
-  if getmetatable(val) == NodeProxy then
-    return tostring(val) .. get_prefixed_addr("userdata: ", val)
-  elseif type(val) == "table" and has_custom_tostring(val) then
-    return tostring(val) .. " " .. get_prefixed_addr("table: ", val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:44]: here we take metatable twice almost in 100% cases (`getmetatable()`, `has_custom_tostring()`)
So it's better to reduce count of such calls

File: engine/engine/content/builtins/scripts/edn.lua [hunk 6]
```diff
@@ -14,34 +20,15 @@ local escape_char_map = {
...
+  local mt = debug.getmetatable(val)
+  if mt then 
+    if mt.__metatable == NodeProxy then
+      return string.format("%s%p", val, val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:27]: `%p` for getting address

File: engine/engine/content/builtins/scripts/edn.lua [hunk 7]
```diff
@@ -68,38 +55,34 @@ local function encode_number(val)
...
 
 local type_to_primitive_encoder = {
   ["nil"] = encode_nil,
   ["boolean"] = tostring,
   ["number"] = encode_number,
   ["string"] = encode_string,
-  ["table"] = make_ref_encoder(get_table_ref),
-  ["userdata"] = make_tagged_string_encoder("#lua/userdata"),
-  ["function"] = make_tagged_string_encoder("#lua/function"),
-  ["thread"] = make_tagged_string_encoder("#lua/thread"),
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:102]: functional-like code looks nicer, but on a big count of operations it works slower because adds extra scopes should be kept and extra function calls
+ generalisation of everything adds a few heavycalls in places where it's not necessary (see comment in `encode_function`)

File: engine/engine/content/builtins/scripts/edn.lua [hunk 8]
```diff
@@ -68,38 +55,34 @@ local function encode_number(val)
...
-    return encode_as_tagged_string(tag, val)
-  end
+function encode_userdata(val)
+  return '#lua/userdata"'..str(val)..'"'
 end
 
-local function get_table_ref(t)
-  return get_prefixed_addr("table: ", t)
+function encode_function(val)
+  return '#lua/function"'..str(val)..'"'
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:70]: We know how function serialised tostring so, it's not necessary to call heavy `encode_string` function here

The same for `encode_thread` but I'm not sure about `encode_userdata`, what do you think?

File: engine/engine/content/builtins/scripts/edn.lua [hunk 9]
```diff
@@ -111,57 +94,60 @@ local function encode_as_primitive(val)
...
+      dups[val] = true
+      refs[#refs + 1] = val
+      local next_depth = depth + 1
+      for k, v in pairs(val) do
+        collect_refs(next_depth, k, refs, dups)
+        collect_refs(next_depth, v, refs, dups)
+      end
+    end
+  end
+end
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:127]: The same function, just another place.

I tried to implement `flat` function without recurtion, but it's slower because of more garbage and GC.
So I keep it as is

File: engine/engine/content/builtins/scripts/edn.lua [hunk 10]
```diff
@@ -111,57 +94,60 @@ local function encode_as_primitive(val)
...
-  return refs
-end
-
 local function encode_table(val)
   local encoded_kvs = {}
   for k, v in pairs(val) do
-    table.insert(
-      encoded_kvs,
-      encode_as_primitive(k) .. " " .. encode_as_primitive(v))
+    encoded_kvs[#encoded_kvs + 1] = encode_as_primitive(k) .. " " .. encode_as_primitive(v)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:100]: on big count of values difference in speed between `table.insert()` and `encoded_kvs[#encoded_kvs + 1]` well noticable

File: engine/engine/content/builtins/scripts/edn.lua [hunk 11]
```diff
@@ -111,57 +94,60 @@ local function encode_as_primitive(val)
...
 local function encode_refs(refs)
   local encoded_refs = {}
-  for k, v in pairs(refs) do
-    table.insert(
-      encoded_refs,
-      encode_as_primitive(k) .. " " .. encode_table(v))
+  local val
+  for i = 1, #refs do
+    val = refs[i]
+    encoded_refs[i] = encode_as_primitive(val).." "..encode_table(val)
```
Comment (AGulev) [inline engine/engine/content/builtins/scripts/edn.lua:110]: Here we are talking about 450 000 - 500 000 values, on such numbers `fori` MUCH faster then `forp`
To be able to use it I added a separate table `dups` for duplicates for `collect_refs`

File: engine/engine/content/builtins/scripts/edn.lua (hunk 1 above)
Comment (britzl) [reply engine/engine/content/builtins/scripts/edn.lua:13]: I like it! Do we want to take it further and also store local copies of string.format and string.gsub?

File: engine/engine/content/builtins/scripts/edn.lua [hunk 12]
```diff
@@ -68,38 +55,34 @@ local function encode_number(val)
 end
 
 local function encode_string(val)
-  return '"' .. val:gsub('[%z\1-\31\\"]', escape_char) .. '"'
+  return '"'..val:gsub('[%z\1-\31\\"]', escape_char)..'"'
```
Comment (britzl) [inline engine/engine/content/builtins/scripts/edn.lua:58]: Do we ever call encode_string() on the same string multiple times? If so, perhaps cache the result and trade memory for performance?

File: engine/engine/content/builtins/scripts/edn.lua (hunk 1 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua:13]: I tried it and it was Slower O_o
it was deep night, maybe I missunderstood something or maybe some kind of LuaJit optimisaitons, not sure, let me try again

File: engine/engine/content/builtins/scripts/edn.lua (hunk 11 above)
Comment (britzl) [reply engine/engine/content/builtins/scripts/edn.lua:110]: I don't understand why we are calling both encode_as_primitive() and encode_table() on `val`?

The old code did `encode_as_primitive(k)` on the table key. What is the key with this change?

**Review by britzl — CHANGES_REQUESTED** [review]
Overall some nice improvements. I have some questions though!

File: engine/engine/content/builtins/scripts/edn.lua (hunk 11 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua:110]: yes, but old table was
refs[val] = val
so the key was equal to the value

File: engine/engine/content/builtins/scripts/edn.lua (hunk 1 above)
Comment (britzl) [reply engine/engine/content/builtins/scripts/edn.lua:13]: Huh, that's a surprise then! Perhaps there's some kind of limit on optimization of locals if they are too many?

File: engine/engine/content/builtins/scripts/edn.lua (hunk 11 above)
Comment (britzl) [reply engine/engine/content/builtins/scripts/edn.lua:110]: Oh, ok, weird? But then I guess it is fine.

File: engine/engine/content/builtins/scripts/edn.lua (hunk 12 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua:58]: Great Idea!
With strings cache:
```
DEBUG:SCRIPT: Result time = 	5.7142062187195	 cached_string	2056917	 encoded_strs	476642
-- and if I don't clean cache (which make sence because I can just step over) it is:
DEBUG:SCRIPT: Result time = 	5.2303631305695	 cached_string	2533328	 encoded_strs	68
DEBUG:SCRIPT: Result time = 	5.1561901569366	 cached_string	2533404	 encoded_strs	40
DEBUG:SCRIPT: Result time = 	5.0541839599609	 cached_string	2533514	 encoded_strs	60
DEBUG:SCRIPT: Result time = 	5.0240840911865	 cached_string	2533549	 encoded_strs	34
```
I'll add some limit when I'll clean cache, but it works good.
Also it gives food for thought about too much work we does all the time

File: engine/engine/content/builtins/scripts/edn.lua (hunk 1 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua:13]: yeah, it's strange, but new tests shows the same result. The diff is minimum but it's a bit worse 🤷‍♂️

File: engine/engine/content/builtins/scripts/edn.lua (hunk 12 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua]: I added strings cache

File: engine/engine/content/builtins/scripts/edn.lua [hunk 13]
```diff
@@ -111,57 +102,61 @@ local function encode_as_primitive(val)
...
 
 local function encode_refs(refs)
   local encoded_refs = {}
-  for k, v in pairs(refs) do
-    table.insert(
-      encoded_refs,
-      encode_as_primitive(k) .. " " .. encode_table(v))
+  local val
+  for i = 1, #refs do
+    val = refs[i]
```
Comment (vlaaad) [inline engine/engine/content/builtins/scripts/edn.lua:117]: Just curious, why not `local val = refs[i]`?

**Review by vlaaad — APPROVED** [review]

File: engine/engine/content/builtins/scripts/edn.lua (hunk 13 above)
Comment (AGulev) [reply engine/engine/content/builtins/scripts/edn.lua:117]: no hidden sence, just an old habbit
in this case there is no difference at all

**Review by britzl — APPROVED** [review]
